### ISSUE TITLE
Caching images for less bandwidth usage

### ISSUE THREAD

**Issue by ellsclytn** [IgnisDa/ryot#263]
I noticed that a pretty large amount of data can be downloaded when using Ryot. This appears to be due to most images being downloaded at their `original` resolution (at least, this is the case for Tmdb images. I haven't investigated other media sources in as much detail yet).

I could have a go at making a PR to address this, but I first wanted to pose the problem in an Issue so I could get opinions on ways of reaching a solution.

So far I've been thinking that this could be solved by changing the API slightly so that instead of `image` properties being a string, they would be an object describing the various various scales available. E.g.

```ts
type ScaledImage = {
    scale: string;
    url: string;
};

type ImageScales = {
    default: string;
    alternative_scales: ScaledImage[];
};
```

An example response of this in use via the GraphQL API might look something like:

```json
{
	"1": {
		"identifier": "65",
		"lot": "SHOW",
		"title": "Futurama",
		"image": {
			"default": "https://image.tmdb.org/t/p/w185/7RRHbCUtAsVmKI6FEMzZB6Re88P.jpg",
			"alternative_scales": [
				{
					"scale": "2x",
					"url": "https://image.tmdb.org/t/p/w500/7RRHbCUtAsVmKI6FEMzZB6Re88P.jpg"
				},
				{
					"scale": "3x",
					"url": "https://image.tmdb.org/t/p/w780/7RRHbCUtAsVmKI6FEMzZB6Re88P.jpg"
				},
			]
		},
		"publishYear": 1999
	}
}
```

The front end could leverage this by using [using `srcset` on the `img` elements](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#resolution_switching_same_size_different_resolutions). It would, however, be a breaking change to the API if it was done exactly as above (maybe it could be put under a different field to `image`).

I've done a little bit of poking at the codebase, so I would be happy to attempt this, but keen to hear your thoughts on the idea first. Thanks.

**Comment by IgnisDa** [comment]
Yes, this was actually raised in https://github.com/IgnisDa/ryot/issues/243. 

The problem is that images are not cached properly in the frontend due to a faulty service worker. I do not have any prior experience with PWAs, so my attempts to fix the issue have not worked.

Here is the cache entry that is responsible for caching images in the frontend, but it does not work:
https://github.com/IgnisDa/ryot/blob/bf5df9b612e90c9bcb9f44ac5e5f98fce4298e09/apps/frontend/worker/cache.js#L7-L17

I would appreciate help there.

**Comment by IgnisDa** [comment]
![image](https://github.com/IgnisDa/ryot/assets/60938164/43a162b5-6453-4d37-8d4d-7cb72ca19104)

As you can see from this screenshot, a lot of stuff is cached, but could never get images to be cached.

**Comment by ellsclytn** [comment]
I guess I assumed the browser would naturally handle image caching to some degree. I was coming at it more from an angle of loading smaller images where practical to reduce unnecessary data transfer in the first place (particularly to help reduce data transfer when potentially operating on cellular/low data allowance connections).

I might read into service workers a little, but haven't done any development with them personally so I probably can't help much in that regard, sorry.

**Comment by IgnisDa** [comment]
Serving smaller images will definitely help, but it will just be a temporary fix to a much larger problem of a faulty service worker. I would rather that the caching problem is fixed first, after which your solution can be implemented if needed.

**Comment by ellsclytn** [comment]
Looks like you beat me to it 😉 nice work, 2.7.0 shows image caching working well on my machine. I'll keep an eye on bandwidth usage while using Ryot for a while. Hopefully your caching work is enough to keep data transfer at a reasonably low level.

Thanks again 🙂

**Comment by IgnisDa** [comment]
Yes bandwidth usage was something that was bugging me for the past few months. Glad that I finally figured it out. Thanks for creating this issue. It pushed me to ask my question on Reddit and they guided me enough that I figured it out.

**Comment by ellsclytn** [comment]
Not sure if this is worthy of a new issue or not, but I noticed the bandwidth is still somewhat high in terms of initial image download. After viewing three shows, I'd transferred over 100MB. While reducing the image sizes would certainly make a difference, the thing I noticed this time is that all carousel images get downloaded. There can be quite a few images for some shows, and they may never get viewed. 

I'm thinking that it might be possible to lazy load those images (or maybe only load one image ahead on the carousel). I don't know exactly what's possible in that regard. I'll take a little look at how the carousel works sometime in the coming days, but wanted to get your thoughts on the issue nonetheless.
